src/find_sort_location.rs

```diff
@@ -0,0 +1,142 @@
+}
+
+impl<F: Field, R: Ring> SecureSort<'_, R, F> {
+    pub async fn execute(
+        self,
+        shares: &[ReplicatedSecretSharing<F>],
+    ) -> Vec<ReplicatedSecretSharing<F>> {
+        let gen_permutation_share = |mult_output: Vec<ReplicatedSecretSharing<F>>| {
+            let len = mult_output.len() / 2;
+
+            // Fold the result
+            let mut permutation_share = Vec::new();
+            (0..len).for_each(|j| {
+                permutation_share.push(mult_output[j] + mult_output[j + len]);
+            });
+            permutation_share
+        };
+        gen_permutation_share(self.secure_multiply(self.prepare_mult_inputs(shares)).await)
+    }
+
+    fn prepare_mult_inputs(
+        &self,
+        shares: &[ReplicatedSecretSharing<F>],
+    ) -> Vec<(ReplicatedSecretSharing<F>, ReplicatedSecretSharing<F>)> {
+        // Step1: Calculate x and 1 - x
+        let mut x_and_one_minus_x: Vec<ReplicatedSecretSharing<F>> =
+            shares.iter().map(|x| self.share_of_one - *x).collect();
+        x_and_one_minus_x.append(&mut shares.to_owned());
+
+        // Step2: Calculate cumulative sum
+        let length = x_and_one_minus_x.len();
+        let mut cumulative_sum = Vec::with_capacity(length);
+        cumulative_sum.push(x_and_one_minus_x[0]);
+        for i in 1..length {
+            cumulative_sum.push(cumulative_sum[i - 1] + x_and_one_minus_x[i]);
+        }
+
+        // Step3: Consolidate step 1 and step 2 output
+        x_and_one_minus_x
+            .iter()
+            .zip(cumulative_sum.iter())
+            .map(|(x_and_one_minus, cumulative_sum)| (*x_and_one_minus, *cumulative_sum))
+            .collect()
+    }
+
+    async fn secure_multiply(
+        &self,
+        mult_inputs: Vec<(ReplicatedSecretSharing<F>, ReplicatedSecretSharing<F>)>,
+    ) -> Vec<ReplicatedSecretSharing<F>> {
+        // Spawn all multiplication, wait for them to finish in parallel and then collect the results
+        let async_multiply = mult_inputs.iter().enumerate().map(|(index, input)| {
+            let output = SecureMul {
+                a_share: input.0,
+                b_share: input.1,
+                index: index as u128,
+            };
+            output.execute(self.context)
+        });
+        join_all(
+            async_multiply
+                .into_iter()
+                .map(|handle| async { handle.await.unwrap() }),
+        )
+        .await
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        field::Fp31,
+        helpers::{self, ring::mock::TestHelper},
+        replicated_secret_sharing::ReplicatedSecretSharing,
+    };
+
+    use super::SecureSort;
+
+    #[tokio::test]
+    async fn find_sorted_position() {
+        let ring = helpers::ring::mock::make_three();
+        let participants = crate::prss::test::make_three();
+        let context = crate::securemul::tests::make_context(&ring, &participants);
+        let mut secure_sort_output = Vec::new();
```

Comment:
```suggestion
        let mut secure_sort_output = Vec::with_capacity(input.len());
```

tho this would require moving `let input` above this line